In [1]:
!pip install pymystem3 -U

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Step-1.-Open-the-data-file-and-examine-the-general-information" data-toc-modified-id="Step-1.-Open-the-data-file-and-examine-the-general-information-1">Step 1. Open the data file and examine the general information</a></span></li><li><span><a href="#Step-2.-Data-preprocessing" data-toc-modified-id="Step-2.-Data-preprocessing-2">Step 2. Data preprocessing</a></span><ul class="toc-item"><li><span><a href="#2.1-Missing-data-processing" data-toc-modified-id="2.1-Missing-data-processing-2.1">2.1 Missing data processing</a></span></li><li><span><a href="#2.2-Replacing-the-data-type" data-toc-modified-id="2.2-Replacing-the-data-type-2.2">2.2 Replacing the data type</a></span></li><li><span><a href="#2.3-Handling-duplicates" data-toc-modified-id="2.3-Handling-duplicates-2.3">2.3 Handling duplicates</a></span></li><li><span><a href="#2.4-Lemmatization" data-toc-modified-id="2.4-Lemmatization-2.4">2.4 Lemmatization</a></span></li><li><span><a href="#2.5-Data-categorization" data-toc-modified-id="2.5-Data-categorization-2.5">2.5 Data categorization</a></span></li></ul></li><li><span><a href="#Step-3:-Answer-the-questions" data-toc-modified-id="Step-3:-Answer-the-questions-3">Step 3: Answer the questions</a></span></li><li><span><a href="#Step-4.-General-conclusion" data-toc-modified-id="Step-4.-General-conclusion-4">Step 4. General conclusion</a></span></li></ul></div>

# Borrower reliability study

Analyze and answer the following questions:   
Is there a dependence between having children and repaying a loan on time?  
Is there a dependence between family status and loan repayment on time?  
Is there a dependence between income level and loan repayment on time?  
How do different purposes of a loan affect its repayment on time?   


The customer is the credit department of the bank. Input data from the bank - statistics on the solvency of customers.


Data Description  

number of children in the family   
-total work experience in days   
-customer age
-the level of education of the client   
-education level identifier   
-family status 
-family status identifier   
-type of employment   
-loan replayment delays   
-monthly income   
-purpose of a loan 


## Step 1. Open the data file and examine the general information

In [2]:
import pandas as pd
from collections import Counter
from pymystem3 import Mystem
import warnings
warnings.simplefilter("ignore") # import the library

data = pd.read_csv('~/datasets/data(1).csv') # save to variable

data.info() # display summary information about the table

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**Conclusion**

The table has 12 columns and 21525 rows and contains integer, float, object and string values

## Step 2. Data preprocessing



###  2.1 Missing data processing

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

The table contains missing values in the days_employed and total_income columns and the wrong data type in the days_employed column

In [4]:
data.isna().sum() # count missing values

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

fill in the gaps in the column days_employed with the median

In [5]:
print('gaps:', data['days_employed'].isna().sum()) # print the number of gaps up to

data.loc[data['days_employed'].isna(), 'days_employed'] = data['days_employed'].median() # fill in the median using the boolean indexing method

print('gaps:', data['days_employed'].isna().sum()) # now print the number of gaps after

gaps: 2174
gaps: 0


fill in the gaps with the median in the income_type column depending on the employment group of the clients

In [6]:
#look at unique values in a column

display(data['income_type'].unique())

array(['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете'],
      dtype=object)

In [7]:
print('Skip:', data['total_income'].isna().sum()) # print the number of gaps up to


medians = data.groupby('income_type')['total_income'].median()
print(medians)

# use the logical indexing method to replace the missing values in each category
data.loc[data['total_income'].isna() & data['income_type'] == 'безработный', 'total_income'] = medians['безработный']

data.loc[data['total_income'].isna() & data['income_type'] == 'в декрете', 'total_income'] = medians['в декрете']
data.loc[(data['total_income'].isna()) & (data['income_type'] == 'госслужащий'), 'total_income'] = medians['госслужащий']
data.loc[(data['total_income'].isna()) & (data['income_type'] == 'компаньон'), 'total_income'] = medians['компаньон']
data.loc[(data['total_income'].isna()) & (data['income_type'] == 'пенсионер'), 'total_income'] = medians['пенсионер']
data.loc[(data['total_income'].isna()) & (data['income_type'] == 'предприниматель'), 'total_income'] = medians['предприниматель']
data.loc[(data['total_income'].isna()) & (data['income_type'] == 'сотрудник'), 'total_income'] = medians['сотрудник']
data.loc[data['total_income'].isna() & data['income_type'] == 'студент', 'total_income'] = medians['студент']

print('Skip:', data['total_income'].isna().sum()) # print the number of gaps 

Skip: 2174
income_type
безработный        131339.751676
в декрете           53829.130729
госслужащий        150447.935283
компаньон          172357.950966
пенсионер          118514.486412
предприниматель    499163.144947
сотрудник          142594.396847
студент             98201.625314
Name: total_income, dtype: float64
Skip: 0


We have filled in all the gaps in the data and now we can move on to the next step

### 2.2 Replacing the data type

We got rid of the gaps, now let's move on to replacing the data type in the 'days_employed' column since the values do not correspond to reality, perhaps the data was taken from two systems, one counts in hours and the other in days

In [8]:
data['days_employed'] = data['days_employed'].apply(abs) # apply the apply(abs) method to make the data positive

display(data.head(10))

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   СРЕДНЕЕ             1   
8         2    6929.865299         35    ВЫСШЕЕ             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Abnormally large values in 'days_employed', which are given in hours, we will translate into days, first we determine in which line the failure occurred and what value we will take as a guide.
To do this, we use the sort_values method.

In [9]:
dt2 = data.sort_values(by = 'days_employed', ascending = True) #sort the table in ascending order

display(dt2.head(10)) # display the first 10 lines

children  days_employed  dob_years education  education_id  \
17437         1      24.141633         31   среднее             1   
8336          0      24.240695         32    высшее             0   
6157          2      30.195337         47   среднее             1   
9683          0      33.520665         43   среднее             1   
2127          1      34.701045         31    высшее             0   
5287          1      37.726602         36   среднее             1   
17270         0      39.954170         34    высшее             0   
13846         2      46.952793         33   среднее             1   
7964          0      47.109840         49    высшее             0   
3235          0      50.128298         43   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
17437        женат / замужем                 0      F    сотрудник     1   
8336   Не женат / не замужем                 4      M    сотрудник     0   
6157        гражданский брак                 1      M    компаньон     0   
9683   Не женат / не замужем                 4      M    сотрудник     1   
2127         женат / замужем                 0      F    компаньон     0   
5287         женат / замужем                 0      F    сотрудник     0   
17270        женат / замужем                 0      F  госслужащий     0   
13846        женат / замужем                 0      F    сотрудник     1   
7964         женат / замужем                 0      F    сотрудник     0   
3235        гражданский брак                 1      F    компаньон     0   

        total_income                                purpose  
17437  166952.415427                     высшее образование  
8336   124115.373655  получение дополнительного образования  
6157   231461.185606                                свадьба  
9683   128555.897209                приобретение автомобиля  
2127    90557.994311                  получение образования  
5287    64954.565099        операции со своей недвижимостью  
17270   97264.767002                                  жилье  
13846  177554.195351                покупка жилья для сдачи  
7964   197545.271278     на покупку подержанного автомобиля  
3235    99381.947946                  на проведение свадьбы

Then, after analyzing the sorted table in parts using the iloc() function, I came to the conclusion that the anomaly occurred between lines 18079 and 18081.

In [10]:
display(dt2.iloc[18079:18081]) 

children  days_employed  dob_years education  education_id  \
16335         1   18388.949901         61   среднее             1   
20444         0  328728.720605         72   среднее             1   

         family_status  family_status_id gender income_type  debt  \
16335  женат / замужем                 0      F   сотрудник     0   
20444   вдовец / вдова                 2      F   пенсионер     0   

        total_income                   purpose  
16335  186178.934089  операции с недвижимостью  
20444   96519.339647   покупка жилья для семьи

Thus, we can take the value of 20,000 as a guideline. Convert hours to days using the logical indexing method.

In [11]:
data.loc[data['days_employed'] > 20000, 'days_employed'] = data.loc[data['days_employed'] > 20000, 'days_employed']/24 

display(data.head(10)) 

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0   14177.753002         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   СРЕДНЕЕ             1   
8         2    6929.865299         35    ВЫСШЕЕ             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Now cast the 'days_employed' column and total_income to the integer type int.

In [12]:
data['days_employed'] = data['days_employed'].astype(int)

data['total_income'] = data['total_income'].astype(int) 

data.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   days_employed     21525 non-null  int32 
 2   dob_years         21525 non-null  int64 
 3   education         21525 non-null  object
 4   education_id      21525 non-null  int64 
 5   family_status     21525 non-null  object
 6   family_status_id  21525 non-null  int64 
 7   gender            21525 non-null  object
 8   income_type       21525 non-null  object
 9   debt              21525 non-null  int64 
 10  total_income      21525 non-null  int32 
 11  purpose           21525 non-null  object
dtypes: int32(2), int64(5), object(5)
memory usage: 1.8+ MB


In this section, we've fixed the data glitch(bug) and converted the data in the days_employed column to a single type.

### 2.3 Handling duplicates

In this stage, I'll find out how many explicit duplicates are in the table. To do this, I will use the duplicated() method in combination with the sum() method and apply it to the entire table.

In [13]:
data.duplicated().sum() #count the number of rows - duplicates

data = data.drop_duplicates().reset_index(drop=True) # save the table cleared of duplicates in the data variable

### 2.4 Lemmatization

In this section, I'll handle lemmatization. I'll highlight the lemmas in the values of the column with the purposes of obtaining a loan . To do this, we use the pymystem3 library

In [14]:
 #import the library

m = Mystem()

text = data['purpose'].unique() #finding unique values (words)
print(text)

joined_text = " ".join(text) #glue them into text

lemmas = m.lemmatize(joined_text) #lemmatize
print(lemmas)

['покупка жилья' 'приобретение автомобиля' 'дополнительное образование'
 'сыграть свадьбу' 'операции с жильем' 'образование'
 'на проведение свадьбы' 'покупка жилья для семьи' 'покупка недвижимости'
 'покупка коммерческой недвижимости' 'покупка жилой недвижимости'
 'строительство собственной недвижимости' 'недвижимость'
 'строительство недвижимости' 'на покупку подержанного автомобиля'
 'на покупку своего автомобиля' 'операции с коммерческой недвижимостью'
 'строительство жилой недвижимости' 'жилье'
 'операции со своей недвижимостью' 'автомобили' 'заняться образованием'
 'сделка с подержанным автомобилем' 'получение образования' 'автомобиль'
 'свадьба' 'получение дополнительного образования' 'покупка своего жилья'
 'операции с недвижимостью' 'получение высшего образования'
 'свой автомобиль' 'сделка с автомобилем' 'профильное образование'
 'высшее образование' 'покупка жилья для сдачи' 'на покупку автомобиля'
 'ремонт жилью' 'заняться высшим образованием']
['покупка', ' ', 'жилье', ' '

Using the Counter, we will calculate the frequency of occurrence of words and determine the most common nouns, they will be the main categories (goals for obtaining a loan)

In [15]:
print(Counter(lemmas))


Counter({' ': 96, 'покупка': 10, 'недвижимость': 10, 'автомобиль': 9, 'образование': 9, 'жилье': 7, 'с': 5, 'операция': 4, 'на': 4, 'свой': 4, 'свадьба': 3, 'строительство': 3, 'получение': 3, 'высокий': 3, 'дополнительный': 2, 'для': 2, 'коммерческий': 2, 'жилой': 2, 'подержать': 2, 'заниматься': 2, 'сделка': 2, 'приобретение': 1, 'сыграть': 1, 'проведение': 1, 'семья': 1, 'собственный': 1, 'со': 1, 'профильный': 1, 'сдача': 1, 'ремонт': 1, '\n': 1})


In [16]:
def purpose_lemmas(row): # write a classification function and apply it to the 'purpose' column
    
    purpose = row['purpose']
    lemmas_purpose = m.lemmatize(purpose)
    
    if ('недвижимость' in lemmas_purpose  or 'жилье' in lemmas_purpose):
        return 'недвижимость'
    elif 'автомобиль' in lemmas_purpose:
        return 'автомобиль'
    elif 'образование' in lemmas_purpose:
        return 'образование'
    if 'свадьба' in lemmas_purpose:
        return 'свадьба'
    else:
        return 'цель не определена'



In [ ]:
data['credit_purpose'] = data.apply(purpose_lemmas, axis=1) # Add a new column with values-results of the function purpose_lemmas()
display(data.head())


**Conclusion** 

Lemmatization of the purposes of the loan is complete. At this point, a new column was added to our dataframe.

### 2.5 Data categorization

In this section, I'll categorize the data and see what data is stored in each column and how it can be classified.
Starting with lowercasing the education column.

In [ ]:
# To begin, let's bring the education column to lowercase.
data['education'] = data['education'].str.lower() 


After analyzing the data table, I can conclude that separate dictionaries are created for different categories of data.
I'll create a new table for the education and education_id columns because they are from the same category. First, let's display all the unique values of the columns

In [ ]:
print(data['education'].value_counts()) 

print(data['education_id'].value_counts())

In [ ]:
education_dict =data[['education', 'education_id']] # create a new 'dictionary'
print(education_dict.head(10)) #print the first 10 lines of the dictionary

There are a lot of duplicates in the 'dictionary', we'll remove them.

In [ ]:
education_dict = education_dict.drop_duplicates().set_index('education_id') #sets the 'education_id' column as an index for the dataframe
display(education_dict.head())#display the first 5 lines

Next, I'll create a dictionary for the family status category. Let's create a new table for the family_status and family_status_id columns.

In [ ]:
family_dict = data[['family_status', 'family_status_id']]
family_dict = family_dict.drop_duplicates().set_index('family_status_id')
display(family_dict.head()) 

Next, we research the children column in the analysis, it will come in handy. Let's display 2 categories with children and without children. To do this, we use a function.

In [ ]:
data['children'] = data['children'].apply(abs) # before classifying, let's bring the data in the column to absolute values, perhaps this column, like the others, had anomalous values.

def children_group(children): # write a classification function and apply it to the 'children' column
     if children == 0:
        return 'no children'
    
     return 'there are children'

data['children_group'] = data['children'].apply(children_group) #create a separate column with categories
display (data.head(10)) # display our dataframe

By analogy with the children column, We'll categorize the dob_years column by writing a function.

In [ ]:
def age_group(dob_years):
     
    if dob_years <= 25 :
         return 'to 25'
    if dob_years > 25 and dob_years <= 50:
         return 'from 25 to 50'
    return 'above 50'

print(age_group(18)) #check our function
print(age_group(45))
print(age_group(72))

# I decided not to create a separate column for age_group, since it won't be useful for further research.

By analogy with the previous columns, we will categorize in the total_income column.

In [ ]:
print(data['total_income'].min()) #display the minimum value

print(data['total_income'].max()) #and the maximum

Knowing the minimum and maximum values makes it easier to categorize data into groups.
Now let's write a function and display the data in a separate column.

In [ ]:
def total_income_group(total_income):
     
    if total_income <= 30000 :
         return 'up to 30000'
    if total_income > 30000 and total_income <= 100000:
         return 'from 30000 to 100000'
    if total_income > 100000 and total_income <= 500000:
         return 'from 100000 to 500000'
    
     return 'above 500000'

data['total_income_group'] = data['total_income'].apply(total_income_group) #create a separate column with categories
display (data.head(10)) # display our dataframe

Сonclusion : Two dictionaries education_dict and family_dict were created, the education column was lowercase, and the original table was expanded to include two columns that will be useful for answering questions in the next paragraph.

## Step 3: Answer the questions

- Is there a relationship between having children and repaying the loan on time?

To answer the question, I'll make a pivot table and display a pie chart

In [ ]:
children_debt = data.pivot_table(index=['children_group'], values='debt', aggfunc=['sum', 'count', 'mean']) 
#make a pivot table(where index is the column or columns by which group data, 
#values - values by which we want to see the pivot table, aggfunc - functions applied to values)
children_debt.columns=['debt', 'total','%']
children_debt.sort_values(by= ['total'], ascending = False)
#sort the table in descending order
children_debt['%'] = (children_debt['%'] * 100).round(2) #convert ratio to percentage and round values
display(children_debt)
children_debt.plot.pie(y = '%', figsize=(10, 10))#draw a pie chart

According to the data, ~7.5% of the entire group "without children" are debtors, and ~9.2% of the entire group "with children" are also debtors.

- Is there a connection between family status and loan repayment on time?


By analogy with the previous task, I will  make a pivot table and a pie chart.

In [ ]:
family_debt = data.pivot_table(index=['family_status','family_status_id'], values='debt', aggfunc=['sum', 'count', 'mean'])
family_debt.columns=['debt', 'total','%']
family_debt.sort_values(by= ['total'], ascending = False)
family_debt['%'] = (family_debt['%'] * 100).round(2)
display (family_debt)

family_debt.plot.pie(y = '%', figsize=(10, 10))

Based on the data, the main part of non-payers are people who are not married - up 9.7% and in a civil marriage - 9.3%, the category of people in an official marriage - 7.5%, divorced -7.1% and widowers - 6.6%

- Is there a connection between the level of income and repayment of the loan on time?


In [ ]:
total_income_debt = data.pivot_table(index=['total_income_group'], values='debt', aggfunc=['sum', 'count', 'mean'])
total_income_debt.columns=['debt', 'total','%']
total_income_debt.sort_values(by= ['total'], ascending = False)
total_income_debt['%'] = (total_income_debt['%'] * 100).round(2)
display (total_income_debt)
total_income_debt.plot.pie(y = '%', figsize=(10, 10))

I can say that the category of people whose income is less than 30,000 are rather unreliable borrowers, in contrast to the group of people whose income is over 500,000. There is a relationship between income and loan repayment.

- How do different purposes of the loan affect its repayment on time?

In [ ]:
purpose_debt = data.pivot_table(index=['credit_purpose'], values='debt', aggfunc=['sum', 'count', 'mean'])
purpose_debt.columns=['debt', 'total','%']
purpose_debt.sort_values(by= ['total'], ascending = False)
purpose_debt['%'] = (purpose_debt['%'] * 100).round(2)
display (purpose_debt)
purpose_debt.plot.pie(y = '%', figsize=(10, 10))

Based on the information provided, the following statement can be made - car loans and education loans are the most vulnerable categories, the share of debt on them is almost identical

## Step 4. General conclusion

I have done a lot of work in the field of data analysis based on the statistics of the bank's customers.
After analyzing the data, I can say that the category of bank customers 'with children' has a higher percentage of loan defaults, and also, people who are not officially married have a higher percentage of loan defaults, unlike other categories. I also want to note that citizens with higher incomes have better return rates than the category of citizens with incomes below 30,000 and, surprisingly, people who take out a mortgage loan are more responsible payers than, for example, car loan holders.